components/modules/net.c

```diff
@@ -586,40 +609,44 @@ int net_send( lua_State *L ) {
     luaL_unref(L, LUA_REGISTRYINDEX, ud->client.cb_sent_ref);
     ud->client.cb_sent_ref = luaL_ref(L, LUA_REGISTRYINDEX);
   }
-  if (ud->type == TYPE_UDP_SOCKET && !ud->pcb) {
-    ud->udp_pcb = udp_new();
-    if (!ud->udp_pcb)
-      return luaL_error(L, "cannot allocate PCB");
-    udp_recv(ud->udp_pcb, net_udp_recv_cb, ud);
-    ip_addr_t laddr = IPADDR_ANY_TYPE_INIT;
-    err_t err = udp_bind(ud->udp_pcb, &laddr, 0);
+
+  if (ud->type == TYPE_UDP_SOCKET && !ud->netconn) {
+    ud->netconn = netconn_new_with_callback(NETCONN_UDP, lnet_netconn_callback);
+    if (!ud->netconn)
+      return luaL_error(L, "cannot allocate netconn");
+    ud->netconn->socket = (int)ud;
+    err_t err = netconn_bind(ud->netconn, IP_ADDR_ANY, 0);
     if (err != ERR_OK) {
-      udp_remove(ud->udp_pcb);
-      ud->udp_pcb = NULL;
+      NETCONN_CLOSE(ud->netconn);
+      ud->netconn = NULL;
       return lwip_lua_checkerr(L, err);
     }
     if (ud->self_ref == LUA_NOREF) {
       lua_pushvalue(L, 1);
       ud->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
     }
   }
-  if (!ud->pcb || ud->self_ref == LUA_NOREF)
+
+  if (!ud->netconn || ud->self_ref == LUA_NOREF)
     return luaL_error(L, "not connected");
+
   err_t err;
   if (ud->type == TYPE_UDP_SOCKET) {
-    struct pbuf *pb = pbuf_alloc(PBUF_TRANSPORT, datalen, PBUF_RAM);
-    if (!pb)
+    struct netbuf *buf = netbuf_new();
+    if (!buf || !netbuf_alloc(buf, datalen))
       return luaL_error(L, "cannot allocate message buffer");
-    pbuf_take(pb, data, datalen);
-    err = udp_sendto(ud->udp_pcb, pb, &addr, port);
-    pbuf_free(pb);
+    netbuf_take(buf, data, datalen);
+    err = netconn_sendto(ud->netconn, buf, &addr, port);
+    netbuf_delete(buf);
+
     if (ud->client.cb_sent_ref != LUA_NOREF) {
       lua_rawgeti(L, LUA_REGISTRYINDEX, ud->client.cb_sent_ref);
       lua_rawgeti(L, LUA_REGISTRYINDEX, ud->self_ref);
       lua_call(L, 1, 0);
     }
   } else if (ud->type == TYPE_TCP_CLIENT) {
-    err = tcp_write(ud->tcp_pcb, data, datalen, TCP_WRITE_FLAG_COPY);
+    size_t bytes_written;
+    err = netconn_write_partly(ud->netconn, data, datalen, NETCONN_COPY, &bytes_written);
```

Comment:
Presumably `bytes_written` should be checked and we should do something if we can't send it all?